### PR TITLE
EIP-7002:Execution layer triggerable withdrawals

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

 Feature

**What does this PR do? Why is it needed?**

implements https://eips.ethereum.org/EIPS/eip-7002 with changes described in https://github.com/ethereum/consensus-specs/blob/dev/specs/electra/beacon-chain.md

- [x] includes spec tests

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

### CODE REVIEW

File: beacon-chain/core/electra/transition.go [hunk 1]
```diff
@@ -27,6 +27,9 @@ var (
 	ProcessSyncCommitteeUpdates          = altair.ProcessSyncCommitteeUpdates
 	AttestationsDelta                    = altair.AttestationsDelta
 	ProcessSyncAggregate                 = altair.ProcessSyncAggregate
+	// These need to be rewritten for electra.
+	ProcessDeposits                      = altair.ProcessDeposits
+	ProcessAttestationsNoVerifySignature = altair.ProcessAttestationsNoVerifySignature
```
Comment (terencechain) [inline beacon-chain/core/electra/transition.go:32]: If these are not related to EIP7002, we probably shouldn't have them, even as placeholders it's a little confusing to the reader. Instead, we could open issues to track them if really needed, which is cleaner....

File: beacon-chain/core/electra/withdrawals.go [hunk 2]
```diff
@@ -75,6 +85,103 @@ import (
...
+		}
+		// # Verify withdrawal credentials
+		hasCorrectCredential := helpers.HasExecutionWithdrawalCredentials(validator)
+		isCorrectSourceAddress := bytes.Equal(validator.WithdrawalCredentials[12:], wr.SourceAddress)
+		if !hasCorrectCredential || !isCorrectSourceAddress {
+			continue
+		}
+
+		currentSlot := st.Slot()
+		currentEpoch := slots.ToEpoch(currentSlot)
```
Comment (terencechain) [inline beacon-chain/core/electra/withdrawals.go:120]: Could be defined outside of the for loop

File: beacon-chain/core/electra/withdrawals.go [hunk 3]
```diff
@@ -75,6 +85,103 @@ import (
...
+		}
+
+		pendingBalanceToWithdraw, err := st.PendingBalanceToWithdraw(vIdx)
+		if err != nil {
+			return nil, err
+		}
+		if isFullExitRequest {
+			//# Only exit validator if it has no pending withdrawals in the queue
+			if pendingBalanceToWithdraw == 0 {
+				maxExitEpoch, churn := validators.MaxExitEpochAndChurn(st)
```
Comment (terencechain) [inline beacon-chain/core/electra/withdrawals.go:142]: Could be defined outside of the for loop

File: beacon-chain/core/electra/withdrawals.go [hunk 4]
```diff
@@ -75,6 +85,103 @@ import (
...
+			return nil, errors.New("nil execution layer withdrawal request")
+		}
+		amount := wr.Amount
+		isFullExitRequest := amount == params.BeaconConfig().FullExitRequestAmount
+		// If partial withdrawal queue is full, only full exits are processed
+		if n, err := st.NumPendingPartialWithdrawals(); err != nil {
+			return nil, err
+		} else if n == params.BeaconConfig().PendingPartialWithdrawalsLimit && !isFullExitRequest {
+			// if the PendingPartialWithdrawalsLimit is met, the user would have paid for a partial withdrawal that's not included
+			continue
```
Comment (terencechain) [inline beacon-chain/core/electra/withdrawals.go:101]: I'm not sure if this condition needs to be covered by the unit test

File: beacon-chain/core/electra/withdrawals.go [hunk 5]
```diff
@@ -75,6 +85,103 @@ import (
...
+		}
+
+		vIdx, exists := st.ValidatorIndexByPubkey(bytesutil.ToBytes48(wr.ValidatorPubkey))
+		if !exists {
+			continue
```
Comment (terencechain) [inline beacon-chain/core/electra/withdrawals.go:106]: I wonder if it's worth adding a debug log at any of these continue statements. I'm not too worried about verbosity or spam. If it continues, it means something was willing to waste some gas, which may be useful for debugging.

File: beacon-chain/core/electra/transition.go (hunk 1 above)
Comment (james-prysm) [reply beacon-chain/core/electra/transition.go:32]: will remove

File: beacon-chain/core/electra/withdrawals.go (hunk 4 above)
Comment (james-prysm) [reply beacon-chain/core/electra/withdrawals.go:101]: can try to add something

File: beacon-chain/core/electra/withdrawals.go (hunk 3 above)
Comment (james-prysm) [reply beacon-chain/core/electra/withdrawals.go:142]: I don't think so, if there are multiple values the loop updates the state I think

Comment (terencechain) [reply beacon-chain/core/electra/withdrawals.go:142]: Ah. Good point

File: beacon-chain/core/electra/withdrawals.go [hunk 6]
```diff
@@ -75,6 +86,107 @@ import (
...
+		vIdx, exists := st.ValidatorIndexByPubkey(bytesutil.ToBytes48(wr.ValidatorPubkey))
+		if !exists {
+			log.Debugln("Skipping execution layer withdrawal request, validator index not found")
+			continue
+		}
+		validator, err := st.ValidatorAtIndex(vIdx)
+		if err != nil {
+			return nil, err
+		}
+		// # Verify withdrawal credentials
```
Comment (saolyn) [inline beacon-chain/core/electra/withdrawals.go:116]: Nits
```suggestion
		// Verify withdrawal credentials
```

File: beacon-chain/core/electra/withdrawals.go [hunk 7]
```diff
@@ -75,6 +86,107 @@ import (
...
+		hasCorrectCredential := helpers.HasExecutionWithdrawalCredentials(validator)
+		isCorrectSourceAddress := bytes.Equal(validator.WithdrawalCredentials[12:], wr.SourceAddress)
+		if !hasCorrectCredential || !isCorrectSourceAddress {
+			log.Debugln("Skipping execution layer withdrawal request, wrong withdrawals credentials")
```
Comment (saolyn) [inline beacon-chain/core/electra/withdrawals.go:120]:
```suggestion
			log.Debugln("Skipping execution layer withdrawal request, wrong withdrawal credentials")
```

File: beacon-chain/core/electra/withdrawals.go [hunk 8]
```diff
@@ -75,6 +86,107 @@ import (
...
+			log.Debugln("Skipping execution layer withdrawal request, validator has not been active long enough")
+			continue
+		}
+
+		pendingBalanceToWithdraw, err := st.PendingBalanceToWithdraw(vIdx)
+		if err != nil {
+			return nil, err
+		}
+		if isFullExitRequest {
+			//# Only exit validator if it has no pending withdrawals in the queue
```
Comment (saolyn) [inline beacon-chain/core/electra/withdrawals.go:145]:
```suggestion
			// Only exit validator if it has no pending withdrawals in the queue
```

File: beacon-chain/core/electra/withdrawals.go [hunk 9]
```diff
@@ -75,6 +86,107 @@ import (
...
+			return nil, err
+		} else if n == params.BeaconConfig().PendingPartialWithdrawalsLimit && !isFullExitRequest {
+			// if the PendingPartialWithdrawalsLimit is met, the user would have paid for a partial withdrawal that's not included
+			log.Debugln("Skipping execution layer withdrawal request, PendingPartialWithdrawalsLimit reached")
+			continue
+		}
+
+		vIdx, exists := st.ValidatorIndexByPubkey(bytesutil.ToBytes48(wr.ValidatorPubkey))
+		if !exists {
+			log.Debugln("Skipping execution layer withdrawal request, validator index not found")
```
Comment (saolyn) [inline beacon-chain/core/electra/withdrawals.go:109]: Would it be helpful to add the `bytesutil.ToBytes48(wr.ValidatorPubkey)` to the log so that it's clear which validator is the issue?

File: beacon-chain/core/electra/withdrawals.go [hunk 10]
```diff
@@ -75,6 +86,107 @@ import (
...
+			return nil, err
+		}
+		hasExcessBalance := vBal > params.BeaconConfig().MinActivationBalance+pendingBalanceToWithdraw
+
+		// Only allow partial withdrawals with compounding withdrawal credentials
+		if helpers.HasCompoundingWithdrawalCredential(validator) && hasSufficientEffectiveBalance && hasExcessBalance {
+			//to_withdraw = min(
+			//	state.balances[index] - MIN_ACTIVATION_BALANCE - pending_balance_to_withdraw,
+			//	amount
+			//)
```
Comment (rkapka) [inline beacon-chain/core/electra/withdrawals.go:169]: Can you precede this with a comment saying that this code is from the spec? Something like:
`// Spec definition:`

File: beacon-chain/core/electra/withdrawals.go [hunk 11]
```diff
@@ -75,6 +86,107 @@ import (
...
+			// note: you can safely subtract these values because haxExcessBalance is checked
+			toWithdraw := min(vBal-params.BeaconConfig().MinActivationBalance-pendingBalanceToWithdraw, amount)
+			exitQueueEpoch, err := st.ExitEpochAndUpdateChurn(primitives.Gwei(toWithdraw))
+			if err != nil {
+				return nil, err
+			}
+			// safe add the uint64 to avoid overflow
+			withdrawableEpoch, err := exitQueueEpoch.SafeAddEpoch(params.BeaconConfig().MinValidatorWithdrawabilityDelay)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to add withdrawablility delay to exit queue")
```
Comment (rkapka) [inline beacon-chain/core/electra/withdrawals.go:180]:
```suggestion
				return nil, errors.Wrap(err, "failed to add withdrawability delay to exit queue epoch")
```

File: beacon-chain/core/electra/withdrawals.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (james-prysm) [reply beacon-chain/core/electra/withdrawals.go:109]: hmm not sure i used debugln since i know it adds a new line not sure if debug also does

File: beacon-chain/core/electra/withdrawals.go (hunk 4 above)
Comment (james-prysm) [reply beacon-chain/core/electra/withdrawals.go]: added a test

**Review by rkapka — APPROVED** [review]
